File: 63_unique_path_II/uniquePathsWithObstables.py
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        length = len(obstacleGrid[0])
        width = len(obstacleGrid)
        
        #print obstacleGrid
        obstacleGrid[0][0] = 0 if obstacleGrid[0][0] == 1 else 1
        
        for i in range(1, length):
            if obstacleGrid[0][i] == 1:
                obstacleGrid[0][i] = 0
            else:
                obstacleGrid[0][i] = obstacleGrid[0][i-1]

        for j in range(1, width):
            if obstacleGrid[j][0] == 1:
                obstacleGrid[j][0] = 0
            else:
                obstacleGrid[j][0] = obstacleGrid[j-1][0]
        
        #print obstacleGrid 
       
        for i in range(1, width):
            for j in range(1, length):
                if obstacleGrid[i][j] == 1:
                    obstacleGrid[i][j] = 0
                else:
                    obstacleGrid[i][j] = obstacleGrid[i-1][j] + obstacleGrid[i][j-1]

        #print obstacleGrid
        return obstacleGrid[width-1][length-1]


if __name__ == '__main__':
    sol = Solution()
    obstacleGrid = [[1, 0]]
    print sol.uniquePathsWithObstacles(obstacleGrid)

File: 63_unique_path_II/uniquePathsWithObstables.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid)
    {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        vector<vector<int> > paths(m, vector<int>(n, 0));
        paths[0][0] = obstacleGrid[0][0] == 0? 1 : 0;
        for(int j = 1; j < n; j++)
            if(obstacleGrid[0][j] != 1)
                paths[0][j] = paths[0][j - 1];

        for(int i = 1; i < m; i++)
            if(obstacleGrid[i][0] != 1)
                paths[i][0] = paths[i - 1][0];

        for(int i = 1; i < m; i++)
            for(int j = 1; j < n; j++)
            {
                if(obstacleGrid[i][j] == 0)
                    paths[i][j] = paths[i - 1][j] + paths[i][j - 1];
                else
                    paths[i][j] = 0;
            }

        return paths[m - 1][n - 1];
    }
};